iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp

```diff
@@ -220,6 +210,25 @@ void setDestroyFlagAndClearContainer(vector& container)
     container.clear();
 }
 
+TEST_F(PortManager_test, DoDiscoveryWithInvalidServiceDescriptionLeadsToTermination)
+{
+    PublisherOptions publisherOptions{1U, iox::NodeName_t("node"), false};
+
+    PublisherPortUser publisher(
+        m_portManager
+            ->acquirePublisherPortData(
+                {iox::capro::InvalidIdString, iox::capro::InvalidIdString, iox::capro::InvalidIdString},
+                publisherOptions,
+                "guiseppe",
+                m_payloadDataSegmentMemoryManager,
+                PortConfigInfo())
+            .value());
+    ASSERT_TRUE(publisher);
+    publisher.offer();
+
+    EXPECT_DEATH({ m_portManager->doDiscovery(); }, ".*");
```

Comment:
As we're testing `cxx::E`We should revise all death tests, once we implement the release/debug mode behaviour of `cxx::Expects`.

Reply:
This cannot really happen, can it? We will no request a port with invalid `ServiceDescription` and initialize a publisher with it. Whether discovery should terminate then is debatable, it could just do nothing as well since it cannot offer or subscribe to such ports. But this does not really matter, we should have the system invariant of not providing or managing ports with invalid `ServiceDescription`.

Reply:
Agreed. This code is under our control an therefore we should have only valid `ServiceDescriptions`. In case of an IO boundary, a sanity check has to be performed. Maybe we need two types of `ServiceDescritions`, a user facing one which can be invalid and one which cannot e.g. `ValidatedServiceDescription` which will be created out of the user facing one. We must make more use of the type system to reduce the number of runtime checks.

Reply:
TBH I don't like and disagree with statements "This code is under control". With the next refactoring in v3.0 other developers might work on this part of the code and there might be the case where erroneous code slips in and an invalid will make it to the `PortManager`. My goal when developing safety software is always "Better safe than sorry". Additionally, we need to test **all** possible branches due to MC/DC anyway. Hence, I reject the changes and instead we can allow users to disable checking `cxx::Expects` during runtime e.g. in Release mode.

Reply:
Just to be clear. I didn't suggest to remove the check in the `PortManager`. This is an invariant which needs to be upheld since it is theoretically possible to have an invalid `ServiceDescription`. This is only because the handling with an `cxx::optional` is quite cumbersome in this part of the user interface. But this is again a perfect example why we shouldn't have nullable types. We have to expect them to be null everywhere they are used.

But this test still does not make sense. There should never be a case where `doDiscovery` accesses an invalid `ServiceDescription` since it is not a parameter of the method. It must already be caught in `acquirePublisherPortData` since this is the place where the invariant is broken. This is what I meant with `the code is under our control`.

Agree with you that such a split of a potentially invalid user-facing `ServiceDescription` and a proven always valid internal `ServiceDescription` would be something for another PR.

Oh, and you need to code a bit in Rust. It will make you appreciate the type system and offers you new perspectives to runtime checks ;)

Reply:
Understood. Adapted the test accordingly.

> Oh, and you need to code a bit in Rust. It will make you appreciate the type system and offers you new perspectives to runtime checks ;)

For sure :crab: